lib/internal/Magento/Framework/Xml/Parser.php

```diff
@@ -111,7 +111,10 @@ protected function _xmlToArray($currentNode = false)
                         $value = ['_value' => $value, '_attribute' => $attributes];
                     }
                     if (isset($content[$node->nodeName])) {
-                        if (!isset($content[$node->nodeName][0]) || !is_array($content[$node->nodeName][0])) {
+                        if (
+                            (is_string($content[$node->nodeName]) || !isset($content[$node->nodeName][0]))
+                            || (is_array($value) && !is_array($content[$node->nodeName][0]))
+                        ) {
```

Comment:
Would be really nice to cover new introduced behavior with unit test.

Class implementation looked scary even before this change and every next fixed use case may break previous fixes if there is no enforcement with automated tests.